specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1098 @@
+    - If the `CrossMsgMeta` for the destination exists in the checkpoint, the SCA gets from the `CrossMsgsRegistry` the current `CID` included in the `CrossMsgMeta`, and it appends the newly created cross-net message to `Msgs`. `SCA` then updates with the new `CID` (after appending the message) to the `CrossMsgMeta` for the checkpoint, deletes the outdated `CrossMsgs` from the registry, and includes the updated one.
+    - In these updates, also the total amount of native tokens included in the messages of the `CrossMsgMeta` is updated in the `Value` field.
+- Finally, when the signing window for the checkpoint closes, the checkpoint is propagated including a link to the cross-net message in the `CrossMsgMeta` of the checkpoint. 
+
+> Bottom-up messages and checkpoint propagation data structures
+```go
+
+// Metadata pointing to the list of messages being
+// propagated in the checkpoint.
+type CrossMsgMeta struct {
+    // Source of the cross-net messages included in
+    // this metadata package
+    From    address.SubnetID
+    // Destination of the cross-net messages included.
+    To      address.SubnetID
+    // CID of the aggregation of all the messages being
+    // propagated. This CID is used to uniquely identify
+    // this package of messages.
+    MsgsCid cid.Cid
+    // Nonce of the crossMsgMeta. It is used for partial
+    // ordering and to prevent replay attacks.
+    Nonce   uint64
+    // Aggregation of all the native tokens being transacted
+    // in the included messages.
+    Value   abi.TokenAmount
+}
+
+// CrossMsgs is the data structure used to persist in the
+// `CrossMsgsRegistry` the `Msgs` and `CrossMsgMeta` 
+// propagated in checkpoints
+type CrossMsgs struct {
+    // Raw msgs from the subnet
+    Msgs  []Message   
+    // Metas propagated from child subnets
+    // (full definition of CrossMsgMeta in cross-net message
+    // section)
+    Metas []CrossMsgMeta 
+}
+
+// MetaTag is a convenient struct
+// used to compute the Cid of the MsgMeta
+type MetaTag struct {
+    MsgsCid  cid.Cid
+    MetasCid cid.Cid
+}
+```
+
+#### Executing bottom-up messages
+When a new checkpoint for a child subnet is committed in a network, the `SCA` checks if it includes any `CrossMsgMeta` before storing it in its state. If this is the case, it means that there are pending cross-msgs to be executed or propagated further in the hierarchy. For every `CrossMsgMeta` in the checkpoint, the `SCA`:
+- Checks if the `Value` included in the `CrossMsgMeta` for the source subnet is below the total `CircSupply` of native tokens for the subnet. If the `Value > CircSupply`, `SCA` rejects the cross-msgs included in the `CrossMsgMeta` due to a violation of the firewall requirement. If `Value <= CircSupply` then the `CrossMsgMeta` is accepted, and `CircSupply` is decremented by `Value`.
```

Comment:
I understand how `Fund` and `Release` work because it's clear who will get tokens in which subnet. 

My question is, what happens if Alice in `/root/subnet-1/subnet-11` wants to send 10 FIL to Bob in `/root/subnet-2/subnet-21`? Do cross messages have a notion of the `Value` included in token transfers between users at all? Or is it only the HC related messages where `Value` is tracked?

Let's say that HC was able to cater for this use case. 

If Alice wanted to send from `/root` to `/root/subnet-2/subnet-21`, she would spend 10 in root, get 10 in subnet-2, then she would I assume have to pay the fees for the funds to go into subnet-21. If she doesn't have an account in subnet-2, maybe she'd only be able to forward 10-fees. Bob would get whatever's left.

Going up from `/root/subnet-1/subnet-11` to `/root` would mean the release of her funds, for which she pays the fee in subnet-11, she gets 10 in subnet-1. Then if she didn't have any funds there, she has to pay fees for it to go further up. 


Do I get the fee mechanism right? What if it's not a transfer, when we don't now what kind of message we are sending? Does Alice have to have an account in each subnet along the path to pay for the next leg?

Reply:
I _think_ what you're asking is tentatively explained at the end of the document:

> With this information, cross-net message will include an array of `gas_limit` amounts that the message is willing to allocate to each subnet in the hop. Subnets in the path won't be able to charge over the specified `gas_limit` for the subnet. When a cross-net reaches a subnet for which `gas_limit` is insufficient, it fails and an error message is propagated back to the source, along with the unspent gas.

I don't think Value is affected at all.

Reply:
Great points, @aakoshh. The design for fee handling hasn't been fully finalized. The current proposal is that the `Value` doesn't change in-route, and the source address in the source subnet needs to provision in the message `value + sum(gas_limit_subnet)`. In each subnet in the path to the destination at most `gas_limit_subnet` will be paid (if not the message fails and is not routed further). At destination, the remaining tokens not used for paying fees is deposited in the originators address in the destination subnet.

Actually, I'll update that section with our current thinking so we can start gathering feedback for it.

Reply:
Oh, thanks @jsoares I completely missed that part on my second read. I did not expect it to be at the bottom of the page, rather than colocated with the explanation of bottom-up and top-down message handling.

Thanks @adlrocha, I think I get it. This means (and I think this would be worth pointing out in the text), in my example this does not mean that Alice has to have an account every step of the way; she is charged the full amount in the source subnet, then in the higher level subnets the `CircSupply` is reduced by `Value + remaining-fees`, and the validator proposing the block in the given subnet chips away at the value, up to the limit of the fees set for their level, before sending the transaction on with an adjusted value. Not sure if Alice can get any refunds for unspent gas.

Reply:
> Oh, thanks @jsoares I completely missed that part on my second read. I did not expect it to be at the bottom of the page, rather than colocated with the explanation of bottom-up and top-down message handling.

Fair point. I will move the section up. The reason why is there is because we are exploring these mechanics as part of our collaboration with CEL, and for completeness we kept everything in the crytpo-econ seection.

>  Alice has to have an account every step of the way; she is charged the full amount in the source subnet, then in the higher level subnets the CircSupply is reduced by Value + remaining-fees, and the validator proposing the block in the given subnet chips away at the value, up to the limit of the fees set for their level, before sending the transaction on with an adjusted value.

Correct

> Not sure if Alice can get any refunds for unspent gas.

She can. The remaining fees are routes with the message to the destination subnet. Once the message executed in the destination subnet, if there are any remaining funds in the message, they are deposited in Alice's address in the destination subnet. Makes sense?

Reply:
Yes, that makes sense. 

Although I would not be surprised if users wanted their funds returned to where they came from, so their applications are simplified and don't have to track funds scattered around all the subnets they ever sent messages to. This could happen any time, whenever there is bandwidth, offered as a free service, included in the price of the original hops. In a way it happens anyway if there is an error, to return the unspent fees and provide feedback. 

Perhaps the feedback mechanism could be applied in the successful case as well, so a sent message will always get an echo with the remaining fees and the outcome of whether it could be delivered and is stable. The stability could be indicated by the applied nonces being included in a checkpoint. That might not exist to day, but may be worth doing. 

I think the [IBC protocol](https://ibcprotocol.org/) also contains acknowledgements going back.